WorkflowSwiftUIExperimental/Sources/ObservableValue.swift

```diff
@@ -0,0 +1,106 @@
+
+        return (observableValue, sink)
+    }
+
+    private init(value: Value, isDuplicate: ((Value, Value) -> Bool)?) {
+        self.internalValue = value
+        self.isDuplicate = isDuplicate
+        self.cancellable = valuePublisher()
+            .dropFirst()
+            .sink { [weak self] newValue in
+                guard let self = self else { return }
+                self.objectWillChange.send()
+                self.internalValue = newValue
+            }
+        // Allows removeDuplicates operator to have the initial value.
+        subject.send(value)
+    }
+
+    //// Scopes the ObservableValue to a subset of Value to LocalValue given the supplied closure while allowing to optionally remove duplicates.
+    /// - Parameters:
+    ///   - toLocalValue: A closure that takes a Value and returns a LocalValue.
+    ///   - isDuplicate: An optional closure that checks to see if a LocalValue is a duplicate.
+    /// - Returns: a scoped ObservableValue of LocalValue.
+    public func scope<LocalValue>(_ toLocalValue: @escaping (Value) -> LocalValue, isDuplicate: ((LocalValue, LocalValue) -> Bool)? = nil) -> ObservableValue<LocalValue> {
+        return scopeToLocalValue(toLocalValue, isDuplicate: isDuplicate)
+    }
+
+    /// Scopes the ObservableValue to a subset of Value to LocalValue given the supplied closure and removes duplicate values using Equatable.
+    /// - Parameter toLocalValue: A closure that takes a Value and returns a LocalValue.
+    /// - Returns: a scoped ObservableValue of LocalValue.
+    public func scope<LocalValue>(_ toLocalValue: @escaping (Value) -> LocalValue) -> ObservableValue<LocalValue> where LocalValue: Equatable {
+        return scopeToLocalValue(toLocalValue, isDuplicate: { $0 == $1 })
+    }
+
+    /// Returns the value at the given keypath of ``Value``.
+    ///
+    /// In combination with `@dynamicMemberLookup`, this allows us to write `model.myProperty` instead of
+    /// `model.value.myProperty` where `model` has type `ObservableValue<T>`.
+    public subscript<T>(dynamicMember keyPath: KeyPath<Value, T>) -> T {
+        internalValue[keyPath: keyPath]
+    }
+
+    private func scopeToLocalValue<LocalValue>(_ toLocalValue: @escaping (Value) -> LocalValue, isDuplicate: ((LocalValue, LocalValue) -> Bool)? = nil) -> ObservableValue<LocalValue> {
+        let localObservableValue = ObservableValue<LocalValue>(
+            value: toLocalValue(internalValue),
+            isDuplicate: isDuplicate
+        )
+        localObservableValue.parentCancellable = valuePublisher().sink(receiveValue: { newValue in
+            localObservableValue.value = toLocalValue(newValue)
+        })
+        return localObservableValue
+    }
+
+    private func valuePublisher() -> AnyPublisher<Value, Never> {
+        guard let isDuplicate = isDuplicate else {
+            return subject.eraseToAnyPublisher()
+        }
+
+        return subject.removeDuplicates(by: isDuplicate).eraseToAnyPublisher()
```

Comment:
If we wrote
```
subject.removeDuplicates(by: self.isEquivalent).eraseToAnyPublisher()
```
then the returned publisher would hold a strong reference to the function `isEquivalent` but not to `self` since the parameter of `removeDuplicates` is not an autoclosure.

I.e., after the line
```
let x = ObservableValue(…).valuePublisher()
```
the `ObservableValue` should be deallocated. Does that sound right to you?

Reply:
ah yes, that makes sense, thanks for walking through it. i was playing around with some sample code as i recall some sort of 'gotcha' when passing instance properties around as closures, and i _believe_ the thing i was thinking of was that if we had something like:

```swift
// method on ObservableValue
func isEquivalent(v1: Value,  v2: Value) -> Bool
```

and then passed _that_ as the filter parameter like

```swift
subject
  .removeDupliates(by: isEquivalent(v1:v2:))...
```

then that pipeline would retain the ObservableValue instance as the 'closure' being passed to the filter is actually an instance method (presumably `self` is an implicit parameter in that formulation, so has to be retained). anyway, file this away as mostly a curio that may occasionally cause surprises.

Reply:
Yes, I think that is exactly right. I have been bitten on multiple occasions by passing an instance method as an escaping function parameter.